evm_arithmetization/src/prover.rs

```diff
@@ -30,132 +30,148 @@ use starky::stark::Stark;
 
 use crate::all_stark::{AllStark, Table, NUM_TABLES};
 use crate::cpu::kernel::aggregator::KERNEL;
-use crate::cpu::kernel::interpreter::{generate_segment, ExtraSegmentData};
+use crate::cpu::kernel::interpreter::{
+    generate_segment, generate_segment_given_data, ExtraSegmentData,
+};
 use crate::generation::state::GenerationState;
 use crate::generation::{generate_traces, GenerationInputs, MemBeforeValues, SegmentData};
 use crate::get_challenges::observe_public_values;
 use crate::memory::segments::Segment;
 use crate::proof::{AllProof, MemCap, PublicValues, RegistersData};
-use crate::witness::memory::MemoryAddress;
+use crate::witness::memory::{MemoryAddress, MemoryState};
 use crate::witness::state::RegistersState;
 
+/// Structure holding the data needed to initialize a segment.
+#[derive(Clone, Default, Debug)]
+pub struct GenerationSegmentData {
+    /// Registers at the start of the segment execution.
+    pub registers: RegistersState,
```

Comment:
does it need to be `pub`?

Reply:
This is so we can have access to the registers when providing `registers_after` as inputs. Butactually, it's probably mich better to have a getter for the registers instead!